File: src/NuclearReactor.cpp
#include "NuclearReactor.h"
vector<Move> moves;
vector<Point> FinalPoints;
//checks whether a bomb/defusal kit can be placed at the given column number
bool NuclearReactor::isBombPlaceable(int column_number){
    if(cells[0][column_number].getStatus() == Constants::EMPTY){
        return true;
    }
    else{
        return false;
    }
}

NuclearReactor::NuclearReactor(NuclearReactor const &NR){
            for(int i = 0; i < Constants::ROWS; i++)
                for(int j = 0; j < Constants::COLUMNS; j++)
                    this->cells[i][j] = NR.cells[i][j];
            this->lastmove = NR.lastmove;
}

//Constructor with the starting player as parameter
NuclearReactor::NuclearReactor(Constants::PlayerType player){
    //cout << "Game Board Created with first play by " << player << endl;
    for(int i = 0; i < Constants::ROWS; i++){
        for(int j = 0; j < Constants::COLUMNS; j++){
            cells[i][j].setRowIndex(i);
            cells[i][j].setColumnIndex(j);
        }
    }
    lastmove.setRowIndex(-1);
    lastmove.setColumnIndex(-1);
    lastmove.setStatus(Constants::EMPTY);
}

//Get the position of the last bomb dropped into the board
NuclearCell NuclearReactor::getRecentOpponentMove(){
    return lastmove;
}

//Get the status of the cell at (row,column)
Constants::NuclearCellStatus NuclearReactor::getCellStatus(int row,int column){
    return cells[row][column].getStatus();
}

//Drops the bomb in the "column" if it is a valid move
void NuclearReactor::dropBomb(int column,Constants::PlayerType P){
    //cout << "Trying to place bomb at " << column << " column by " << player << endl;
    if(isBombPlaceable(column)){
        //cout << "Bomb Placed" << endl;
        int row = Constants::ROWS-1;
        while(cells[row][column].getStatus() != Constants::EMPTY) row--;
        cells[row][column].setStatus((P)?Constants::BOMB : Constants::DEFUSER);
        lastmove.setRowIndex(row);
        lastmove.setColumnIndex(column);
        lastmove.setStatus(cells[row][column].getStatus());
        Move move;
        move.player = ((P)?Constants::userbotname : Constants::compbotname);
        move.row = row;
        move.column = column;
        moves.push_back(move);
    }
    else{
        fstream f(Constants::filename.c_str(),ios::in | ios::out | ios::app);
        writeToJson((P)?Constants::userbotname : Constants::compbotname);
        exit(0);
    }
}

//check whether the game is over
GameState NuclearReactor::isGameOver(){
    //cout << "Checking if game is over" << endl;
    int k,directions[8] = {0};
    bool flags[8] = {false};
    NuclearCell bc = lastmove;
    int row = bc.getRowIndex();
    int column = bc.getColumnIndex();
    Constants::NuclearCellStatus status = bc.getStatus();
    for(k = 0; k < Constants::COLUMNS; k++)
        if(isBombPlaceable(k))
            break;
    if(k == Constants::COLUMNS){
        fstream f(Constants::filename.c_str(),ios::in | ios::out | ios::app);
        f << "],\n\t\"winner\" : \"Draw\"\n}";
        f.close();
        //cout << "End Play" << endl;
        exit(0);
    }
    if(row == -1 && column == -1)
        return NO_STATE;
    for(int i = 0; i < 3; i++){
        for(int j = 0; j < 8; j++){
            if(directions[j] == i && !flags[j]){
                switch(j){
                    case 0:
                        if(row-i-1 >= 0 && row-i-1 < Constants::ROWS  && column-i-1 >=0 && column-i-1 < Constants::COLUMNS){
                            if(getCellStatus(row-i-1,column-i-1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 1:
                        if(row-i-1 >= 0 && row-i-1 < Constants::ROWS){
                            if(getCellStatus(row-i-1,column) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 2:
                        if(row-i-1 >= 0 && row-i-1 < Constants::ROWS && column+i+1 >=0 && column+i+1 < Constants::COLUMNS){
                            if(getCellStatus(row-i-1,column+i+1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 3:
                        if(column+i+1 >=0 && column+i+1 < Constants::COLUMNS){
                            if(getCellStatus(row,column+i+1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 4:
                        if(row+i+1 >= 0 && row+i+1 < Constants::ROWS && column+i+1 >=0 && column+i+1 < Constants::COLUMNS){
                            if(getCellStatus(row+i+1,column+i+1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 5:
                        if(row+i+1 >= 0 && row+i+1 < Constants::ROWS){
                            if(getCellStatus(row+i+1,column) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 6:
                        if(row+i+1 >= 0 && row+i+1 < Constants::ROWS && column-i-1 >=0 && column-i-1 < Constants::COLUMNS){
                            if(getCellStatus(row+i+1,column-i-1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                    case 7:
                        if(column-i-1 >=0 && column-i-1 < Constants::COLUMNS){
                            if(getCellStatus(row,column-i-1) == status){
                                directions[j]++;
                            }
                            else{
                                flags[j] = ~flags[j];
                            }
                        }
                        break;
                }
            }
        }
    }
    Point points[4];
    if((directions[0]+directions[4]) >= 3 || (directions[1]+directions[5]) >= 3 || (directions[2]+directions[6]) >= 3 || (directions[3]+directions[7]) >= 3){
       if((directions[0]+directions[4]) >= 3){
            int i = 0;
            for(int j = 0; i<3 && j < directions[0]; j++){
                points[i].row = row - 1 - j;
                points[i++].column = column - 1 - j;
            }
            for(int j = 0; i<3 && j < directions[4]; j++){
                points[i].row = row + 1 + j;
                points[i++].column = column + 1 + j;
            }
        }
        else if((directions[1]+directions[5]) >= 3){
            int i = 0;
            for(int j = 0; i<3 && j < directions[1]; j++){
                points[i].row = row - 1 - j;
                points[i++].column = column;
            }
            for(int j = 0; i<3 && j < directions[5]; j++){
                points[i].row = row + 1 + j;
                points[i++].column = column;
            }
        }
        else if((directions[2]+directions[6]) >= 3){
            int i = 0;
            for(int j = 0; i<3 && j < directions[2]; j++){
                points[i].row = row - 1 - j;
                points[i++].column = column + 1 + j;
            }
            for(int j = 0; i<3 && j < directions[6]; j++){
                points[i].row = row + 1 + j;
                points[i++].column = column - 1 - j;
            }
        }
        else if((directions[3]+directions[7]) >= 3){
            int i = 0;
            for(int j = 0; i<3 && j < directions[3]; j++){
                points[i].row = row;
                points[i++].column = column + 1 + j;
            }
            for(int j = 0; i<3 && j < directions[7]; j++){
                points[i].row = row;
                points[i++].column = column - 1 - j;
            }
        }
        points[3].row = row;
        points[3].column = column;
        for(int j = 0; j < 4 ; j++){
            FinalPoints.push_back(points[j]);
        }
        return WIN;
    }

    return NO_STATE;
}
void writeToJson(string player){
    cout << "{" << endl;
    cout << "\t\"moves\" : [" << endl;
    for(unsigned int i = 0; i < moves.size(); i++){
        cout << "\t\t{" << endl;
        cout << "\t\t\t\"player\" : \"" << moves[i].player << "\"," << endl;
        cout << "\t\t\t\"row\" : \"" << moves[i].row << "\"," << endl;
        cout << "\t\t\t\"column\" : \"" << moves[i].column << "\"" << endl;
        cout << "\t\t}";
        if(i != moves.size() - 1)
            cout << "," << endl;
        else
            cout << endl;
    }
    cout << "\t]," << endl;

    f << "\t\"Final Points\" : [" << endl;
    for(unsigned int i = 0 ; i < FinalPoints.size(); i++){
        cout << "\t\t{" << endl;
        cout << "\t\t\t\"row\" : \"" << FinalPoints[i].row << "\"," << endl;
        cout << "\t\t\t\"column\" : \"" << FinalPoints[i].column << "\"" << endl;
        cout << "\t\t}";
        if(i != FinalPoints.size() - 1)
            cout << "," << endl;
        else
            cout << endl;
    }
    cout << "\t]" << endl;
    cout << "}";
}

File: Python/NuclearReactor.py
from NuclearCell import NuclearCell
from Constants import Constants
class Move:
    player = ''
    row = 0
    column = 0
    
class Point:
    row = 0
    column = 0

moves = list()
FinalPoints = list()

def writeToJson(player):
    print "{"
    print '\t"moves" : ['
    for move in moves:
        print "\t\t{"
        print '\t\t\t"player" : "' + move.player + '",'
        print '\t\t\t"row" : "' + str(move.row) + '",'
        print '\t\t\t"column" : "' + str(move.column) + '"'
        print "\t\t}" + ("," if(move is not moves[len(moves)-1]) else '')
    
    print "\t],"
    print '\t"coordinates" : ['

    for FinalPoint in FinalPoints:
        print "\t\t{"
        print '\t\t\t"row" : "' + str(FinalPoint.row) + '",'
        print '\t\t\t"column" : "' + str(FinalPoint.column) + '"'
        print "\t\t}" + ("," if(FinalPoint is not FinalPoints[len(FinalPoints) - 1]) else "")

    print"\t],"
    print '\t"winner" : "' + player + '"'
    print "}"
    
class NuclearReactor:
    def __init__(self):
        self.__cells = list()
        for i in range(Constants.ROWS):
            rowcell = list()
            for j in range(Constants.COLUMNS):
                cell = NuclearCell()
                cell.setRowIndex(i)
                cell.setColumnIndex(j)
                rowcell.append(cell)
            self.__cells.append(rowcell)
        self.__lastmove = NuclearCell()

    def getRecentOpponentMove(self):
        return self.__lastmove

    def isBombPlaceable(self,column_number):
        if(self.__cells[0][column_number].getStatus() == Constants.NuclearCellStatus.EMPTY):
            return True
        else:
            return False
        
    def dropBomb(self,column,P):
        if(self.isBombPlaceable(column)):
            row = Constants.ROWS-1
            while(self.__cells[row][column].getStatus() is not Constants.NuclearCellStatus.EMPTY):
                row -= 1
            self.__cells[row][column].setStatus(Constants.NuclearCellStatus.BOMB if P else Constants.NuclearCellStatus.DEFUSER)
            self.__lastmove.setRowIndex(row)
            self.__lastmove.setColumnIndex(column)
            self.__lastmove.setStatus(self.__cells[row][column].getStatus())
            move = Move()
            move.player = Constants.userbotname if P else Constants.compbotname
            move.row = row
            move.column = column
            moves.append(move)
        else:
            #print 'invalid move'
            writeToJson(Constants.userbotname if P else Constants.compbotname)
            exit(0)
       
    def getCellStatus(self,row,column):
        try:
            return self.__cells[row][column].getStatus()
        except:
            return Constants.Constants.NuclearCellStatus.INVALID      

    def isGameOver(self):
        directions = [0]*8
        flags = [False]*8
        n = 0
        row = self.__lastmove.getRowIndex()
        column = self.__lastmove.getColumnIndex()
        status = self.__lastmove.getStatus()

        if( row == -1 and column == -1):
            return Constants.PlayerType.NOPLAYER

        for i in range(3):
            for j in range(8):
                if(directions[j] == i and flags[j] is False):
                    if(j == 0):
                        if(row-i-1 >= 0 and row-i-1 < Constants.ROWS  and column-i-1 >=0 and column-i-1 < Constants.COLUMNS):
                            if(self.getCellStatus(row-i-1,column-i-1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 1):
                        if(row-i-1 >= 0 and row-i-1 < Constants.ROWS):
                            if(self.getCellStatus(row-i-1,column) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 2):
                        if(row-i-1 >= 0 and row-i-1 < Constants.ROWS  and column+i+1 >=0 and column+i+1 < Constants.COLUMNS):
                            if(self.getCellStatus(row-i-1,column+i+1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 3):
                        if(column+i+1 >=0 and column+i+1 < Constants.COLUMNS):
                            if(self.getCellStatus(row,column+i+1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 4):
                        if(row+i+1 >= 0 and row+i+1 < Constants.ROWS  and column+i+1 >=0 and column+i+1 < Constants.COLUMNS):
                            if(self.getCellStatus(row+i+1,column+i+1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 5):
                        if(row+i+1 >= 0 and row+i+1 < Constants.ROWS):
                            if(self.getCellStatus(row+i+1,column) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    elif(j == 6):
                        if(row+i+1 >= 0 and row+i+1 < Constants.ROWS  and column-i-1 >=0 and column-i-1 < Constants.COLUMNS):
                            if(self.getCellStatus(row+i+1,column-i-1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True
                    else:
                        if(column-i-1 >=0 and column-i-1 < Constants.COLUMNS):
                            if(self.getCellStatus(row,column-i-1) is status):
                                directions[j] += 1
                            else:
                                flags[j] = True

        if((directions[0]+directions[4]) >= 3 or (directions[1]+directions[5]) >= 3 or (directions[2]+directions[6]) >= 3 or (directions[3]+directions[7]) >= 3):
            if((directions[0]+directions[4]) >= 3):
                i = 0;
                for j in range(directions[0]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row - 1 - j
                    point.column = column - 1 - j
                    FinalPoints.append(point)
                    i += 1
                for j in range(directions[4]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row + 1 + j
                    point.column = column + 1 + j
                    FinalPoints.append(point)
                    i += 1
            elif((directions[1]+directions[5]) >= 3):
                i = 0;
                for j in range(directions[1]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row - 1 - j
                    point.column = column
                    FinalPoints.append(point)
                    i += 1
                for j in range(directions[5]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row + 1 + j
                    point.column = column
                    FinalPoints.append(point)
                    i += 1
            elif((directions[2]+directions[6]) >= 3):
                i = 0;
                for j in range(directions[2]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row - 1 - j
                    point.column = column + 1 + j
                    FinalPoints.append(point)
                    i += 1
                for j in range(directions[6]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row + 1 + j
                    point.column = column - 1 - j
                    FinalPoints.append(point)
                    i += 1
            else:
                i = 0;
                for j in range(directions[3]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row
                    point.column = column + 1 + j
                    FinalPoints.append(point)
                    i += 1
                for j in range(directions[7]):
                    if(i>=3):
                        break
                    point = Point()
                    point.row = row
                    point.column = column - 1 - j
                    FinalPoints.append(point)
                    i += 1
            point = Point()        
            point.row = row
            point.column = column
            FinalPoints.append(point)
                
            return (Constants.PlayerType.COMP if status else Constants.PlayerType.USER)
        else:
            for k in range(Constants.COLUMNS):
                n += 1
                if(self.isBombPlaceable(k)):
                    break;
            if( n == Constants.COLUMNS):
                print '\t"winner" : "draw"'
                print '}'
                exit(0)
            return Constants.PlayerType.NOPLAYER
